### DISCUSSION TITLE
UKF-JPDAF tracker with Constant velocity in the X and known turn rate for Y and Z keeps diverging

### DISCUSSION THREAD

**Discussion by bkev13** [dstl/Stone-Soup#1088] [Q&A]
I have this implementation of the UKF-JPDAF to track objects defined by their 3D coordinates. My filter performs great in the x-direction, which I represent using the constant velocity model, but it keeps diverging in the y- and z-directions which I represent with the constant turn rate model. I am a bit at a loss as to why.

The filter is supposed to track everything and then identify the load based on some defined conditions and export to a CSV file. In the detections that I provide, everything should be static which makes it even more unusual for it to be diverging in the y and z direction. I am not using the randomwalk model because my implementation should work for scenarios when objects are moving and when they are not. Any assistance will be helpful, it is the last step for me to complete my thesis and I am a bit at a loss.

The script:
[soup_solution_CTR.txt](https://github.com/user-attachments/files/17304612/soup_solution_CTR.txt)

The log_file:
[s1_Test1.csv](https://github.com/user-attachments/files/17304625/s1_Test1.csv)

The output:
[5_track_s1_Test1_target.xlsx](https://github.com/user-attachments/files/17304691/5_track_s1_Test1_target.xlsx)

**Comment by sdhiscocks** [comment]
So looking at the data and your measurement model parameters, it looks to me like the noise value you have is too large. The data is much more distributed on x-axis (from about +15 to +40) vs y- and z- (-5 to +5), so this may be why it appears to be behaving better on x-axis. Try reducing the noise on the measurement model, in particular y- and z-, hopefully with some insight from your sensor.

Also, the KnownTurnRate you have zero noise, which isn't ideal. Setting this to `[target_velocity_noise]*2` would be good place to start.

**Reply by bkev13** [reply]
This feedback really helped thank you. I want to do one last thing, I want to extract the measurement assocaited with every state for a track. I am using this function to extract the states but I can't find the class to call to extract the measurement/detection that is associated with that state. Here is my function:

def state_sequence_to_df(sequences):
    df = pd.DataFrame(
        {'id': sequence.id,
         'date': state.timestamp.date(),
         'time': state.timestamp.strftime('%H:%M:%S.%f')[:-3], # Extract time with milliseconds
         'x': state.state_vector[0],
         'vx': state.state_vector[1],
         'y': state.state_vector[2],
         'vy': state.state_vector[3],
         'z': state.state_vector[4],
         'vz': state.state_vector[5],
        }
        for sequence in sequences
        for state in sequence.last_timestamp_generator()
    )
    return df

**Comment by bkev13** [comment]
Thank you for the reply, I tried implementing your feedback and am getting slightly better results. However, it affects the number of tracks quite a bit. The tracker now creates more sporadic tracks than before, almost double
